### PR TITLE
[build] [tools] Consolidation of directory handling code

### PR DESCRIPTION
**Pull request by ejgallego**
This amounts to moving some of the bits on `Envars` to a new library
`coq.boot`, which is lighter so tools don't have to pull the whole of
`clib` in order to use this.

We take advantage of the move to introduce a stronger API, which
introduces the notion of "Coq runtime enviroment", which can be later
used to access different elements of a Coq install.

In addition to the removal of duplicate and inconsistent code, this
API should make a few things easier, including improvements
w.r.t. re-locatable setups.

We also remove some remnants of the coq-core compatibility layer,
which is not needed after the switch to Dune for the OCaml parts of
Coq.

- https://github.com/LPCIC/coq-elpi/pull/287

### CODE REVIEW

**Comment by ejgallego** [comment]
There is still a bit of work to do, in particular for cleaning up envars stuff, but we can try to get this in already I think.

File: boot/dune
Comment (SkySkimmer) [inline boot/dune:2]: Shouldn't it be fine to depend on `lib`?

File: boot/env.ml
Comment (SkySkimmer) [inline boot/env.ml:65]: IIUC this means if we pass -coqlib then COQCORELIB is ignored, is that really what we want?

File: boot/dune
Comment (ejgallego) [reply boot/dune:2]: There would be no problem, however coqdep / coqdoc etc... depend on this, so it adds a bit of latency in the build.

I'd prefer for now to keep the deps lean, let's see how things look once we factor loadpath out.

File: boot/env.ml
Comment (ejgallego) [reply boot/env.ml:65]: Indeed, rules here are still a bit messy; the plan was to move the handling of `-coqlib` to this lib too, but meanwhile I'll update the PR so if `COQCORELIB` is set, we use it; but it could be the case that we are using `-coqlib` to set both I think.

File: tools/coqdoc/cdglobals.mli
Comment (Lysxia) [inline tools/coqdoc/cdglobals.mli:42]: Is this going to be used?

File: tools/coqdoc/main.ml
Comment (Lysxia) [inline tools/coqdoc/main.ml:397]: This if and the if above seem redundant.

File: tools/coqdoc/cdglobals.mli
Comment (ejgallego) [reply tools/coqdoc/cdglobals.mli:42]: Good catch, this was introduced in a clenaup commit that has been squashed now. Will fix.

File: boot/env.ml
Comment (ejgallego) [reply boot/env.ml:65]: I have changed opinion and I'll try to cleanup this in upcoming PRs, I will properly document what is happening tho.

File: tools/coqdoc/main.ml
Comment (ejgallego) [reply tools/coqdoc/main.ml]: Indeed, thanks.

**Comment by ejgallego** [comment]
@silene I broke this on my last push, but was passing CI since a few weeks; let me fix it.

This is the only PR I'm involved I'd like to get into 8.14, as having a central place for this logic will greatly help if we hit some problem w.r.t. the package split / build system overhaul in 8.14 , which we likely will.

**Comment by ejgallego** [comment]
Should be fixed now, was a silly typo.

File: boot/dune
Comment (ejgallego) [reply boot/dune:2]: Another point in favor of having this more self-contained, is that eventually language servers and build systems would like to vendor this library as to avoid re-doing all the loadpath handling, etc...

File: tools/coqdoc/main.ml
Comment (Zimmi48) [inline tools/coqdoc/main.ml:303]: Maybe you could check here if the argument has a URL format and print an informative error message about the renaming of the option.

Comment (ejgallego) [reply tools/coqdoc/main.ml:315]: Thanks for the suggestion, done.

**Review by Lysxia — APPROVED** [review]
Nothing coqdoc-related to add.

**Comment by ppedrot** [comment]
@ejgallego shouldn't this be demilestoned from 8.14?

File: boot/util.ml
Comment (SkySkimmer) [inline boot/util.ml:1]: No mli for this module?

File: boot/env.mli
Comment (SkySkimmer) [inline boot/env.mli:46]: This comment should be more explicit: we `exit 1` instead of raising an exception.

File: ide/coqide/unicode_bindings.ml
Comment (SkySkimmer) [inline ide/coqide/unicode_bindings.ml:55]: This change means that if coqlib isn't found we `exit 1` instead of printing to console, I don't think that's a good idea.

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
shim is broken:
~~~
dune exec -- dev/shim/coqtop-prelude
Warning:
Cannot open directory /home/gaetan/dev/coq/coq/_build/default/dev/shim/../../../coq-core/plugins
[cannot-open-dir,filesystem]
Welcome to Coq 8.14+alpha
Error:
File not found on loadpath : ltac_plugin.cmxs
Loadpath: /home/gaetan/dev/coq/coq:/home/gaetan/dev/coq/coq/_build/default/user-contrib/Ltac2
~~~

File: boot/util.ml
Comment (ejgallego) [reply boot/util.ml:1]: I marked it internal so I didn't care too much, but indeed we can add it; I expect most of this code to go eventually away.

File: ide/coqide/unicode_bindings.ml
Comment (ejgallego) [reply ide/coqide/unicode_bindings.ml:55]: What do you propose? We could let callers of `Env.init` recover the error, but as far as I can see there is no situation where if the file env is not setup, things would work.

Note also that in this particular case, the call to `Envars.coqlib ()` could return indeed `""` which was even more confusing IMHO, as it would print the message "not found" but don't point out the real cause of the issue [missing call to `coqlib_guess` I think]

**Comment by silene** [comment]
This is too massive a change, so postponing.

**Comment by Alizter** [comment]
Coqdoc cannot be built since it complains about not having the coqlib option. Even when supplied in the doc/dune file, it complains that it could not find the .sty files.

**Comment by ejgallego** [comment]
> The refman cannot be built since coqdoc complains about not having the coqlib option. Even when supplied in the doc/dune file, it complains that it could not find the .sty files. I'm not sure if this is because of the new directory code for coqdoc or the coqlib option being supplied incorrectly.

Thanks a lot @Alizter , this is very helpful

**Comment by Alizter** [comment]
I should mention that previously we didn't need to supply coqlib because coqdoc could guess the directory. It looks like that code was removed however.

**Comment by ejgallego** [comment]
> I should mention that previously we didn't need to supply coqlib because coqdoc could guess the directory. It looks like that code was removed however.

It should still do it, I will look into it; also, I think we want to add some kind of `COQENV_DEBUG` env variable, so we can just set it and get a debug output for the env resolution.

**Comment by Alizter** [comment]
There are some additions to the test-suite in #14718 that are probably relevant here.

File: doc/changelog/08-cli-tools/14059-coq+path_boot.rst
Comment (jfehrle) [inline doc/changelog/08-cli-tools/14059-coq+path_boot.rst:4]:
```suggestion
  to ``--coqlib_url`` and ``--coqlib`` to make them more consistent with
  flags used in other Coq executables
```

(if accurate)

**Review by jfehrle — COMMENTED** [review]
wording tweaks

File: doc/changelog/08-cli-tools/14059-coq+path_boot.rst
Comment (ejgallego) [reply doc/changelog/08-cli-tools/14059-coq+path_boot.rst:4]: Much better, thanks!

**Comment by ejgallego** [comment]
@SkySkimmer the problem you mentioned should be fixed.

This surely needs more refinement, but maybe we should go ahead now and keep working on other PRs?

**Comment by ejgallego** [comment]
In particular, a todo list is:
- force Env is not doubly-initialized
- get rid of set_coqlib and instead pass it as an argument to `init`
- move the loadpath code to boot [that for sure should be done in a different PR]

File: boot/env.mli
Comment (ejgallego) [reply boot/env.mli:46]: Yes, will fix.

File: ide/coqide/unicode_bindings.ml
Comment (ejgallego) [reply ide/coqide/unicode_bindings.ml:55]: I agree tho, we should improve this before merge.

Comment (SkySkimmer) [reply ide/coqide/unicode_bindings.ml:55]: Putting needs:progress for this then

Comment (ejgallego) [reply ide/coqide/unicode_bindings.ml]: See the latest commit, actually, the whole message was bogus to start with, IIUC.

**Comment by ejgallego** [comment]
@SkySkimmer note my comment above [and the removal of the needs:progress tag] ; I'm unsure if it is easy to see with the GH interface hiding stuff.

File: boot/env.mli
Comment (SkySkimmer) [inline boot/env.mli:16]:
```suggestion
This API is similar in spirit to findlib's or dune-sites API,
```

**Review by SkySkimmer — APPROVED** [review]
Just waiting for the typo correction

File: boot/env.mli
Comment (ejgallego) [reply boot/env.mli]: Done thanks.

**Comment by SkySkimmer** [comment]
@coqbot merge now

File: dev/ci/user-overlays/14059-ejgallego-coq+path_boot.sh
Comment (SkySkimmer) [inline dev/ci/user-overlays/14059-ejgallego-coq+path_boot.sh:1]: @ejgallego where's the pr for this?

Comment (ejgallego) [reply dev/ci/user-overlays/14059-ejgallego-coq+path_boot.sh:1]: Ouch, let me open it right now.

Comment (ejgallego) [reply dev/ci/user-overlays/14059-ejgallego-coq+path_boot.sh:1]: https://github.com/LPCIC/coq-elpi/pull/287

**Comment by Zimmi48** [comment]
@ejgallego it seems like you forgot to add code owners for this new library, didn't you?

**Comment by ejgallego** [comment]
> @ejgallego it seems like you forgot to add code owners for this new library, didn't you?

Indeed, thanks for pointing that out.

**Pull request merged**
